Launchpad: Add earn checklist

edanzer opened the pull request:
**UPDATE: I am changing this PR to add a generic Earn checklist, and may add an earn-newsletter checklist later**

## Proposed changes:
* Adds 'earn' checklist to Launchpad. We'll use this to show a Launchpad checklist on Earn pages.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None.

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1) This can be tested via the testing instructions in its companion frontend PR here: 
https://github.com/Automattic/wp-calypso/pull/81930

2) To test in isolation: 
   - Run  to load this branch in your sandbox. 
   - Sandbox public-api and the domain of your test site. 
   - Go to `https://developer.wordpress.com/docs/api/console/`, select `WP REST API` and `wpcom/v2`, and enter the endpoint: `/sites/YOURDOMAIN/launchpad?checklist_slug=earn`
   - Confirm you receive back a valid checklist with the stripe_connected and paid_offer_created tasks.

edanzer committed:
Add earn-newsletter launchpad checklist

edanzer committed:
changelog

valterlorran commented:
Hi @edanzer, would you mind adding a standalone testing instruction here? We can test these changes by requesting `/sites/:siteSlug/launchpad?checklist_slug=earn-newsletter` through the developer console.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -907,6 +915,20 @@ function wpcom_launchpad_is_paid_newsletter_enabled() {
+ */
+function wpcom_launchpad_is_newsletter_or_write_intent() {
+	$intent = get_option( 'site_intent', false );
+	if ( 'newsletter' !== $intent || 'write' !== $intent ) {
```
This condition will always be true. I think we may need `&&` here instead of `||`.
```suggestion
	if ( 'newsletter' !== $intent && 'write' !== $intent ) {
```

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -907,6 +915,20 @@ function wpcom_launchpad_is_paid_newsletter_enabled() {
+ *
+ * @return bool True if the task list is enabled, false otherwise.
+ */
+function wpcom_launchpad_is_newsletter_or_write_intent() {
```
Putting aside the logic issue with the condition in the function, I think we may need to revisit the naming for this function. As far as I can tell, the function should return true when the intent is _not_ `write` or `newsletter`. Can we update the naming so the naming aligns with what we're checking.
```suggestion
function wpcom_launchpad_is_not_newsletter_or_write_intent() {
```

edanzer replied:
@daledupreez The intent is to return true if intent is either write or newsletter (ie, this checklist should be enabled for sites with write/newsletter intent). I do need to adjust though.

```
function wpcom_launchpad_is_newsletter_or_write_intent() {
	$intent = get_option( 'site_intent', false );
	if ( 'newsletter' === $intent || 'write' === $intent ) {
		return true;
	}

	return false;
}
```

I'll update. One odd thing I need to explore is why this has been working as intended when tested alongside my frontend PR.

edanzer replied:
I'm testing this via dev console know. No matter whether I update that condition or not, I get back the same expected checklist, which is why it works. I guess the callback just changes the value of he is_enabled property on the response.

edanzer replied:
Just a note that the method for which this comment was added has been removed because I've updated this PR to be generic, and to show the earn launchpad for all sites, not just newsletter / blog / write.

edanzer committed:
Fix callback conditional

edanzer commented:
@daledupreez @valterlorran -

1) I've added specific instructions for testing via the dev console.

2) I've fixed/updated the conditional for the is_enabled callback. Pretty obvious logic issue. I was surprised when I saw it that it was working at tall. But I think I misunderstood the purpose of the is_enabled callback. Regardless of whether that callback returns true or false, you still get back a checklist with the right tasks. So the frontend PR was working. It's just a matter of whether the is_enabled property on the response root is true or false. Not sure what downstream effect that has, but I was not making use of that property myself. 

Anyways, I believe this PR should be working as expected now. Thanks for the reviews.

valterlorran commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -907,6 +915,20 @@ function wpcom_launchpad_is_paid_newsletter_enabled() {
+ */
+function wpcom_launchpad_is_newsletter_or_write_intent() {
+	$intent = get_option( 'site_intent', false );
+	if ( 'newsletter' === $intent || 'write' === $intent ) {
```
Can we restrict this a little more? I tested with a Free Newsletter and a site with the Write intent, and both returned `enabled = true`, but the checklist was empty. So maybe it shouldn't be enabled when the checklist is empty.

<img width="1149" alt="Screen Shot 2023-09-28 at 17 13 24" src="https://github.com/Automattic/jetpack/assets/1234758/b617064d-1619-46f4-a58f-962464d168f5">

edanzer replied:
> but the checklist was empty

Hmm. Not able to duplicate this. I wouldn't think an empty checklist would be related to the is_enabled callback? If you're seeing an empty checklist, that would seem like it would reflect some other other issue wouldn't it? 

To put it otherwise, if I send a request to `/sites/YOURDOMAIN/launchpad?checklist_slug=earn-newsletter`, given that I'm providing a specific checklist slug, shouldn't I **always** get back the `earn-newsletter` checklist, which is defined to have two items? 

From playing with it, I'd think the is_enabled callback only changes whether the is_enabled property on the request is true or false.

edanzer replied:
To confirm, you'll get 0 checklist items if the site_goals don't include paid_subscribers because the task definitions for set_up_payments and create_paid_newsletter both include `'is_visible_callback' => 'wpcom_launchpad_has_goal_paid_subscribers',`

I need to think how to handle that. I need to be able to load those tasks even if a site doesn't have that goal. 

Options: 
   - Just set a site_goal of paid_newsletters when someone visits the earn page (since they've expressed monetization intent)
   - Create different tasks that don't depend on that is_visible callback.

edanzer replied:
FYI - I recall now that is_visible callback was originally set because we're using the same task definitions for free and paid newsletters and didn't want to show them on paid newsletters. But that intention doesn't travel well at this point, later, when we want to encourage those with free newsletters to monetize. 

It's also a bit odd / non-ideal because I've created a checklist specifically to fetch those two items, but I can't do so without updating the site_goals.

edanzer replied:
I've made several changes based on this and also broader discussion about Launchpad on Earn:
   - I've created two new tasks for for stripe_connected and paid_offer_created that don't have the limitation of needing to have a paid_newsletter goal. 
   - Both tasks above have __return_true as their is_visible callback. For now, I can't think of any reason we would not want those visible on the Earn launchpad.
   - I've updated the checklist slug to be 'earn' instead of 'earn-newsletter'. We had been planning to add a general launchpad to earn after doing it for newsletters. I'm reversing the order and doing the general launchpad first.

edanzer replied:
Just a note that the method for which this comment was added has been removed because I've updated this PR to be generic, and to show the earn launchpad for all sites, not just newsletter / blog / write.

edanzer committed:
Update earn task definitions

edanzer committed:
Make new checklist general

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -183,6 +183,14 @@ function wpcom_launchpad_get_task_list_definitions() {
 			'is_enabled_callback' => 'wpcom_launchpad_is_paid_newsletter_enabled',
 		),
+		'earn'                   => array(
+			'title'               => 'Newsletter',
```
Minor note: this should probably be `Earn` or similar, as it's no longer specific to newsletters.

edanzer replied:
Yup, updated to 'Earn'.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -456,6 +456,32 @@ function wpcom_launchpad_get_task_definitions() {
+		// Earn tasks
+		'stripe_connected'                => array(
+			'get_title'           => function () {
+				return __( 'Connect a Stripe account', 'jetpack-mu-wpcom' );
```
Users may not know what `Stripe` is, or why they might need an account, so I think we may need additional context here. How about `Connect a Stripe account for payments` or `Connect a Stripe account to collect payments`?

daledupreez replied:
Noting that I've shared this feedback on the original designs at p9Jlb4-8NN-p2#comment-9257.

edanzer replied:
Updated this based on the separate P2 discussion.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -456,6 +456,32 @@ function wpcom_launchpad_get_task_definitions() {
+		),
+		'paid_offer_created'              => array(
+			'get_title'           => function () {
+				return __( 'Set up an offer', 'jetpack-mu-wpcom' );
```
I have no idea what an "offer" is or who it might be for. Assuming this is a "payment plan", can we use wording like `Set up a payment plan for [site] visitors`?

daledupreez replied:
Also shared on the original design: p9Jlb4-8NN-p2#comment-9257

edanzer replied:
Updated this and payment text string based on comments here: 
p9Jlb4-8NN-p2#comment-9257

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -456,6 +456,32 @@ function wpcom_launchpad_get_task_definitions() {
+			'is_visible_callback' => '__return_true',
+			'get_calypso_path'    => function ( $task, $default, $data ) {
+				if ( function_exists( 'get_memberships_connected_account_redirect' ) ) {
+					return get_memberships_connected_account_redirect(
```
I followed the link that gets returned here and it looks like we're push users directly to Stripe to kick off the connection workflow. I think we need to make sure we're giving the user more context to make sure they have a good sense of what will happen next when they visit the site, as the change may feel really abrupt.

edanzer replied:
Based on the P2 comments linked elsewhere, we're still going directly to stripe, but have updated the text strings to have greater clarity.

daledupreez reviewed (commented):
This is working as described, but I have some concerns about the UX we'll end up with. Could you include some links to the designs or other context that you're working from so I can share that feedback in the right place?

edanzer commented:
UPDATE: Got it. Additional details on this here: p9Jlb4-8NN-p2#comment-9257

Hey @daledupreez - Here are Figma designs for the earn page: DqXCQr1dEWpF3P2dIEwiwd-fi-791_88014
   - You mentioned 'concerns about the ux' based on this PR... can you clarify what you may be concerned about? 
   - I'll ping and talk to Zap about a) if we might update the the task text for 'stripe' and 'offer' and b) what if anything we might do to provide some intermediate context between the task itself and Stripe.

edanzer committed:
Update task text

edanzer committed:
Updated checklist title to Earn

edanzer commented:
For @daledupreez and others... I've updated the checklist title to 'Earn' and updated the task text strings for greater clarity based on the related conversation at p9Jlb4-8NN-p2#comment-9257. We're still be linking directly to Stripe, but sounds like that's fine given the additional text string clarity.

edanzer committed:
Check if new tasks are complete

edanzer committed:
Remove check for write/newsletter intent

edanzer committed:
Cast values to array

lezama commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+ *
+ * @return bool Whether or not Stripe account is connected.
+ */
+function wpcom_is_stripe_connected() {
```
should we add it to the `lib` instead? sounds useful

daledupreez replied:
Agreed -- this should be `wpcom_launchpad_is_stripe_connected()` if we're keeping it in the Launchpad code, or it should move somewhere else (where "somewhere else" could still be in jetpack-mu-wpcom).
```suggestion
function wpcom_launchpad_is_stripe_connected() {
```

edanzer replied:
Sure. Can handle as a follow up. Your comment made me think I should double check it doesn't already exists in the lib. As far as I can tell, it doesn't, which is surprising.

lezama commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -191,6 +191,14 @@ function wpcom_launchpad_get_task_list_definitions() {
+				'stripe_connected',
+				'paid_offer_created',
+			),
+			'is_enabled_callback' => '__return_true',
```
sure this doesn't make it show up somewhere else?

daledupreez replied:
I think this should be OK -- we don't have any reason a user shouldn't be able to load the task list.

edanzer replied:
No, I don't think there's anyplace where a launchpad checklist is automatically displayed based on this property.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+ * @return bool Whether or not Stripe account is connected.
+ */
+function wpcom_is_stripe_connected() {
+	require_lib( 'memberships' );
```
Is this library defined on Atomic sites or in Jetpack? Assuming it's available via Jetpack, we also need to make sure that the relevant module(s) are enabled.
At a glance, this looks likely to break in that context.

Also, I really dislike `require_lib()` because I don't know what code it is actually looking for in any particular context.

edanzer replied:
Yes, I was just asking about this on slack. The require_lib() method will always exist in simple/atomic environments, but certain libraries may not. Someone thought memberships may not be present by default on atomic, but I'm not sure how that could be possible given how much other code depends on it. 

I don't have strong feelings one way or the other for require_lib(), but it is very widely used in our codebases, so I've assumed it is acceptable, or even preferred, for many. But I'm happy to adjust.

edanzer replied:
@daledupreez - I've updated this so we're only checking membership settings on simple for now. Checking the membership settings on atomic seems like it will require some extra work. 

I'm still using require_lib( 'memberships' ) as it still seems like the most straightforward and common way to pull in the needed methods. Let me know if you have an alternative approach in mind.

daledupreez replied:
Just to clarify, my dislike of `require_lib()` stems from the fact that it effectively masks the underlying dependency for a casual reader, making the dependency somewhat implicit, unless the reader happens to recall exactly how `require_lib()` does lookups. In this particular case, I can't tell what `'memberships'` is, so I can't easily tell what code we're expecting to be present (and is not present, in some cases!).

As a proposed alternative, I always prefer an explicit `require_once`, as it is generally 100% explicit as to what you're trying to pull in. In this specific case, I would go for something like:
```php
	require_once WP_CONTENT_DIR . '/lib/memberships/0-load.php';
```
To me, it would have been way easier to detect a possible problem with the expectations for this more explicit `require_once` approach, and very clear as to how we'd detect whether the files are present. 😁

edanzer replied:
@daledupreez - I do understand your point. I had to do a fair amount of research today to understand how require_lib() works, where it searches for libraries, how it loads them, and when it's safe to use. By comparison, a require statement is simple and direct.

I wasn't quite sure how to include the library and associated methods without it, but you're right that including that 0-load file should do it. That's basically what require_lib() does under the hood. I'd be happy to update in a follow up.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+ * @return bool Whether or not paid membership plans exist.
+ */
+function wpcom_has_paid_membership_plans() {
+	require_lib( 'memberships' );
```
As above for the source of the library/code.

edanzer replied:
https://github.com/Automattic/jetpack/pull/33200#discussion_r1355760498

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+	require_lib( 'memberships' );
+	$blog_id  = get_current_blog_id();
+	$settings = (array) get_memberships_settings_for_site( $blog_id );
+	return boolval( $settings['connected_account_id'] );
```
Will this output a warning if `$settings['connected_account_id']` is undefined? Should we add an `isset()` wrapper to be super-safe?
```suggestion
	return isset( $settings['connected_account_id'] ) && boolval( $settings['connected_account_id'] );
```

edanzer replied:
Good call. Added the extra checks.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+	require_lib( 'memberships' );
+	$blog_id  = get_current_blog_id();
+	$settings = (array) get_memberships_settings_for_site( $blog_id );
+	return count( $settings['products'] ) > 0;
```
Are we guaranteed that `$settings['products']` will be defined and an array? I feel we should be defensive against the data being undefined, especially as `count()` will throw exceptions for null values in newer versions of PHP.
```suggestion
	return isset( $settings['products'] ) && is_array( $settings['products'] ) && ( count( $settings['products'] ) > 0 ); // we could also use `array() !== $settings['products']` for the final check if we want
```

edanzer replied:
Good call. Added the extra checks.

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,30 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+ *
+ * @return bool Whether or not paid membership plans exist.
+ */
+function wpcom_has_paid_membership_plans() {
```
As above RE naming, can we please keep the `wpcom_launchpad_` prefix, or move the code elsewhere?
```suggestion
function wpcom_launchpad_has_paid_membership_plans() {
```

edanzer replied:
Sure! Happy to update.

edanzer replied:
Updated method names.

daledupreez reviewed (commented):
I've noted a few key items as comments.

edanzer committed:
Update methods for membership/plans status

edanzer committed:
Update method names in callbacks

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,52 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+function wpcom_launchpad_get_membership_settings() {
+	$is_atomic = defined( 'IS_ATOMIC' ) && IS_ATOMIC;
+
+	// Memberships lib is only available on Simple sites.
+	// A follow up will fetch membership settings for Atomic.
```
Minor note: the code _and_ the data are only present and directly accessible from WPCOM Simple.

It looks like we should be able to use code very similar to the code block below to fetch this data when we're in an Atomic context. That code is for the `wpcom/v2/memberships/status` endpoint, which will make a proxied API call to WordPress.com when we're not running on WPCOM Simple.
https://github.com/Automattic/jetpack/blob/cce10b2a0c92b53b79bda816e8fcd3899fdb30ad/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/memberships.php#L369-L401

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -855,6 +883,52 @@ function wpcom_launchpad_get_newsletter_subscriber_count() {
+ *
+ * @return array|null Membership settings or null.
+ */
+function wpcom_launchpad_get_membership_settings() {
```
Neatly done in wrapping all this logic and ensuring the code won't explode outside of WPCOM Simple.
